### PR TITLE
Fix migrating values with empty intersection type

### PR DESCRIPTION
**Pull request by turbolent**
Work towards #3096 

## Description

Migrated values may have empty intersection types, as they originally were restricted types, i.e. they have a legacy type.

Creating empty intersection types without (correctly) defensively panics, but prevents migrating values.

Removing the panic and only accepting empty intersection types is not enough: The resulting sema intersection type is then used for various purposes during the migration, e.g. subtype checking when inserting into a dictionary, performing the appropriate transfer (copy or move), etc.

Bring back the legacy type in `sema.IntersectionType`, just like it already exists in `interpreter.StaticIntersectionType`, and restore the subtyping rules which were removed in #2596 (see https://github.com/onflow/cadence/pull/2596/files#diff-c01765841e64b85306258f5ee4daff2bdc5b9e9e107167f55a6dba4162e8bdb8).

The subtyping rules for Cadence 1.0 intersection type got integrated into the restored rules.

Because this is hard to review, I prepared a diff of the original subtyping rules for restricted types (taken from #2596) and the new subtying rules for intersection types of this PR:

```diff
diff --git a/sub-old.go b/sub-new.go
index 576c521b3..492288806 100644
--- a/sub-old.go
+++ b/sub-new.go
@@ -1,10 +1,32 @@
-intersectionSuperType := typedSuperType.Type
+	case *IntersectionType:
+
+		// TODO: replace with
+		//
+		//switch typedSubType := subType.(type) {
+		//case *IntersectionType:
+		//
+		//	// An intersection type `{Us}` is a subtype of an intersection type `{Vs}` / `{Vs}` / `{Vs}`:
+		//	// when `Vs` is a subset of `Us`.
+		//
+		//	return typedSuperType.EffectiveIntersectionSet().
+		//		IsSubsetOf(typedSubType.EffectiveIntersectionSet())
+		//
+		//case *CompositeType:
+		//	// A type `T` is a subtype of an intersection type `{Us}` / `{Us}` / `{Us}`:
+		//	// when `T` conforms to `Us`.
+		//
+		//	return typedSuperType.EffectiveIntersectionSet().
+		//		IsSubsetOf(typedSubType.EffectiveInterfaceConformanceSet())
+		//}
+
+		intersectionSuperType := typedSuperType.LegacyType //nolint:staticcheck
+
 		switch intersectionSuperType {
-		case AnyResourceType, AnyStructType, AnyType:
+		case nil, AnyResourceType, AnyStructType, AnyType:
 
 			switch subType {
 			case AnyResourceType:
-				// `AnyResource` is a subtype of a intersection type
+				// `AnyResource` is a subtype of an intersection type
 				// - `AnyResource{Us}`: not statically;
 				// - `AnyStruct{Us}`: never.
 				// - `Any{Us}`: not statically;
@@ -12,7 +34,7 @@ intersectionSuperType := typedSuperType.Type
 				return false
 
 			case AnyStructType:
-				// `AnyStruct` is a subtype of a intersection type
+				// `AnyStruct` is a subtype of an intersection type
 				// - `AnyStruct{Us}`: not statically.
 				// - `AnyResource{Us}`: never;
 				// - `Any{Us}`: not statically.
@@ -20,7 +42,7 @@ intersectionSuperType := typedSuperType.Type
 				return false
 
 			case AnyType:
-				// `Any` is a subtype of a intersection type
+				// `Any` is a subtype of an intersection type
 				// - `Any{Us}: not statically.`
 				// - `AnyStruct{Us}`: never;
 				// - `AnyResource{Us}`: never;
@@ -31,20 +53,32 @@ intersectionSuperType := typedSuperType.Type
 			switch typedSubType := subType.(type) {
 			case *IntersectionType:
 
-				// A intersection type `T{Us}`
-				// is a subtype of a intersection type `AnyResource{Vs}` / `AnyStruct{Vs}` / `Any{Vs}`:
+				// An intersection type `T{Us}`
+				// is a subtype of an intersection type `AnyResource{Vs}` / `AnyStruct{Vs}` / `Any{Vs}`:
 
-				intersectionSubtype := typedSubType.Type
+				intersectionSubtype := typedSubType.LegacyType //nolint:staticcheck
 				switch intersectionSubtype {
+				case nil:
+					// An intersection type `{Us}` is a subtype of an intersection type `{Vs}` / `{Vs}` / `{Vs}`:
+					// when `Vs` is a subset of `Us`.
+
+					return typedSuperType.EffectiveIntersectionSet().
+						IsSubsetOf(typedSubType.EffectiveIntersectionSet())
+
 				case AnyResourceType, AnyStructType, AnyType:
 					// When `T == AnyResource || T == AnyStruct || T == Any`:
 					// if the intersection type of the subtype
 					// is a subtype of the intersection supertype,
 					// and `Vs` is a subset of `Us`.
 
-					return IsSubType(intersectionSubtype, intersectionSuperType) &&
-						typedSuperType.EffectiveIntersectionSet().
-							IsSubsetOf(typedSubType.EffectiveIntersectionSet())
+					if intersectionSuperType != nil &&
+						!IsSubType(intersectionSubtype, intersectionSuperType) {
+
+						return false
+					}
+
+					return typedSuperType.EffectiveIntersectionSet().
+						IsSubsetOf(typedSubType.EffectiveIntersectionSet())
 				}
 
 				if intersectionSubtype, ok := intersectionSubtype.(*CompositeType); ok {
@@ -54,59 +88,71 @@ intersectionSuperType := typedSuperType.Type
 					// and `T` conforms to `Vs`.
 					// `Us` and `Vs` do *not* have to be subsets.
 
-					return IsSubType(intersectionSubtype, intersectionSuperType) &&
-						typedSuperType.EffectiveIntersectionSet().
-							IsSubsetOf(intersectionSubtype.EffectiveInterfaceConformanceSet())
+					if intersectionSuperType != nil &&
+						!IsSubType(intersectionSubtype, intersectionSuperType) {
+
+						return false
+					}
+
+					return typedSuperType.EffectiveIntersectionSet().
+						IsSubsetOf(intersectionSubtype.EffectiveInterfaceConformanceSet())
 				}
 
 			case *CompositeType:
 				// A type `T`
-				// is a subtype of a intersection type `AnyResource{Us}` / `AnyStruct{Us}` / `Any{Us}`:
+				// is a subtype of an intersection type `AnyResource{Us}` / `AnyStruct{Us}` / `Any{Us}`:
 				// if `T` is a subtype of the intersection supertype,
 				// and `T` conforms to `Us`.
 
-				return IsSubType(typedSubType, typedSuperType.Type) &&
-					typedSuperType.EffectiveIntersectionSet().
-						IsSubsetOf(typedSubType.EffectiveInterfaceConformanceSet())
+				if intersectionSuperType != nil &&
+					!IsSubType(typedSubType, intersectionSuperType) {
+
+					return false
+				}
+
+				return typedSuperType.EffectiveIntersectionSet().
+					IsSubsetOf(typedSubType.EffectiveInterfaceConformanceSet())
 			}
 
 		default:
+			// Supertype (intersection) has a non-Any* legacy type
 
 			switch typedSubType := subType.(type) {
 			case *IntersectionType:
 
-				// A intersection type `T{Us}`
-				// is a subtype of a intersection type `V{Ws}`:
+				// An intersection type `T{Us}`
+				// is a subtype of an intersection type `V{Ws}`:
 
-				switch typedSubType.Type {
-				case AnyResourceType, AnyStructType, AnyType:
+				intersectionSubType := typedSubType.LegacyType //nolint:staticcheck
+				switch intersectionSubType {
+				case nil, AnyResourceType, AnyStructType, AnyType:
 					// When `T == AnyResource || T == AnyStruct || T == Any`:
 					// not statically.
 					return false
 				}
 
-				if intersectionSubType, ok := typedSubType.Type.(*CompositeType); ok {
+				if intersectionSubType, ok := intersectionSubType.(*CompositeType); ok {
 					// When `T != AnyResource && T != AnyStructType && T != Any`: if `T == V`.
 					//
 					// `Us` and `Ws` do *not* have to be subsets:
 					// The owner may freely restrict and unrestrict.
 
-					return intersectionSubType == typedSuperType.Type
+					return intersectionSubType == intersectionSuperType
 				}
 
 			case *CompositeType:
 				// A type `T`
-				// is a subtype of a intersection type `U{Vs}`: if `T <: U`.
+				// is a subtype of an intersection type `U{Vs}`: if `T <: U`.
 				//
 				// The owner may freely restrict.
 
-				return IsSubType(typedSubType, typedSuperType.Type)
+				return IsSubType(typedSubType, intersectionSuperType)
 			}
 
 			switch subType {
 			case AnyResourceType, AnyStructType, AnyType:
 				// A type `T`
-				// is a subtype of a intersection type `AnyResource{Vs}` / `AnyStruct{Vs}` / `Any{Vs}`:
+				// is a subtype of an intersection type `AnyResource{Vs}` / `AnyStruct{Vs}` / `Any{Vs}`:
 				// not statically.
 
 				return false
@@ -114,21 +160,29 @@ intersectionSuperType := typedSuperType.Type
 		}
 
 	case *CompositeType:
+
 		// NOTE: type equality case (composite type `T` is subtype of composite type `U`)
 		// is already handled at beginning of function
+
 		switch typedSubType := subType.(type) {
 		case *IntersectionType:
 
-			// A intersection type `T{Us}`
+			// TODO: bring back once legacy type is removed
+			// An intersection type `{Us}` is never a subtype of a type `V`:
+			//return false
+
+			// TODO: remove support for legacy type
+			// An intersection type `T{Us}`
 			// is a subtype of a type `V`:
 
-			switch typedSubType.Type {
-			case AnyResourceType, AnyStructType, AnyType:
+			legacyType := typedSubType.LegacyType
+			switch legacyType {
+			case nil, AnyResourceType, AnyStructType, AnyType:
 				// When `T == AnyResource || T == AnyStruct || T == Any`: not statically.
 				return false
 			}
 
-			if intersectionSubType, ok := typedSubType.Type.(*CompositeType); ok {
+			if intersectionSubType, ok := legacyType.(*CompositeType); ok {
 				// When `T != AnyResource && T != AnyStruct`: if `T == V`.
 				//
 				// The owner may freely unrestrict.
```

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/sema/type.go [hunk 1]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+		//	// when `T` conforms to `Us`.
+		//
+		//	return typedSuperType.EffectiveIntersectionSet().
+		//		IsSubsetOf(typedSubType.EffectiveInterfaceConformanceSet())
+		//}
+
+		intersectionSuperType := typedSuperType.LegacyType //nolint:staticcheck
+
+		switch intersectionSuperType {
+		case nil, AnyResourceType, AnyStructType, AnyType:
```
Comment (turbolent) [inline runtime/sema/type.go:7685]: Integrated handling for no legacy type, see diff in description

File: runtime/sema/type.go [hunk 2]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+				// is a subtype of an intersection type `AnyResource{Vs}` / `AnyStruct{Vs}` / `Any{Vs}`:
+
+				intersectionSubtype := typedSubType.LegacyType //nolint:staticcheck
+				switch intersectionSubtype {
+				case nil:
+					// An intersection type `{Us}` is a subtype of an intersection type `{Vs}` / `{Vs}` / `{Vs}`:
+					// when `Vs` is a subset of `Us`.
+
+					return typedSuperType.EffectiveIntersectionSet().
+						IsSubsetOf(typedSubType.EffectiveIntersectionSet())
```
Comment (turbolent) [inline runtime/sema/type.go:7726]: Integrated rule for no legacy type. Same as commented out code at top

File: runtime/sema/type.go [hunk 3]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+
+				case AnyResourceType, AnyStructType, AnyType:
+					// When `T == AnyResource || T == AnyStruct || T == Any`:
+					// if the intersection type of the subtype
+					// is a subtype of the intersection supertype,
+					// and `Vs` is a subset of `Us`.
+
+					if intersectionSuperType != nil &&
+						!IsSubType(intersectionSubtype, intersectionSuperType) {
```
Comment (turbolent) [inline runtime/sema/type.go:7735]: Added support for no legacy type

File: runtime/sema/type.go [hunk 4]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+
+				if intersectionSubtype, ok := intersectionSubtype.(*CompositeType); ok {
+					// When `T != AnyResource && T != AnyStruct && T != Any`:
+					// if the intersection type of the subtype
+					// is a subtype of the intersection supertype,
+					// and `T` conforms to `Vs`.
+					// `Us` and `Vs` do *not* have to be subsets.
+
+					if intersectionSuperType != nil &&
+						!IsSubType(intersectionSubtype, intersectionSuperType) {
```
Comment (turbolent) [inline runtime/sema/type.go:7752]: Added support for no legacy type

File: runtime/sema/type.go [hunk 5]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+				}
+
+			case *CompositeType:
+				// A type `T`
+				// is a subtype of an intersection type `AnyResource{Us}` / `AnyStruct{Us}` / `Any{Us}`:
+				// if `T` is a subtype of the intersection supertype,
+				// and `T` conforms to `Us`.
+
+				if intersectionSuperType != nil &&
+					!IsSubType(typedSubType, intersectionSuperType) {
```
Comment (turbolent) [inline runtime/sema/type.go:7768]: Added support for no legacy type

File: runtime/sema/type.go [hunk 6]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+
+			switch typedSubType := subType.(type) {
+			case *IntersectionType:
+
+				// An intersection type `T{Us}`
+				// is a subtype of an intersection type `V{Ws}`:
+
+				intersectionSubType := typedSubType.LegacyType //nolint:staticcheck
+				switch intersectionSubType {
+				case nil, AnyResourceType, AnyStructType, AnyType:
```
Comment (turbolent) [inline runtime/sema/type.go:7788]: Integrated rule for no legacy type

File: runtime/sema/type.go [hunk 7]
```diff
@@ -7659,33 +7659,196 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
...
+					return false
+				}
+
+				if intersectionSubType, ok := intersectionSubType.(*CompositeType); ok {
+					// When `T != AnyResource && T != AnyStructType && T != Any`: if `T == V`.
+					//
+					// `Us` and `Ws` do *not* have to be subsets:
+					// The owner may freely restrict and unrestrict.
+
+					return intersectionSubType == intersectionSuperType
```
Comment (turbolent) [inline runtime/sema/type.go:7800]: Both `intersectionSubType` and `intersectionSuperType` are non-nil here

**Review by dsainati1 — APPROVED** [review]

File: runtime/sema/type.go [hunk 8]
```diff
@@ -7982,12 +8147,15 @@ type IntersectionType struct {
...
+	// Deprecated
+	LegacyType Type
 }
 
 var _ Type = &IntersectionType{}
 
-func NewIntersectionType(memoryGauge common.MemoryGauge, types []*InterfaceType) *IntersectionType {
-	if len(types) == 0 {
+// TODO: remove `legacyType` once all uses of it are removed
+func NewIntersectionType(memoryGauge common.MemoryGauge, legacyType Type, types []*InterfaceType) *IntersectionType {
```
Comment (dsainati1) [inline runtime/sema/type.go:8157]: Personally, I prefer to have the constructor not have a legacy type, and to require setting it manually in the few times it's actually used. IMO this makes it clear that this isn't something to be used in most code, but ultimately it's just a matter of preference I guess.

Comment (turbolent) [reply runtime/sema/type.go:8157]: The legacy type is used to check the validity of the constructed type (`len(types) == 0 && legacyType == nil` is invalid). I'd like to keep the check in place, as it helped us already identify some cases we did not handle properly

**Review by SupunS — APPROVED** [review]

**Pull request merged**